File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    players_dict = {player: rank for rank, player in enumerate(players)}
    players_dict_reverse = {rank: player for rank, player in enumerate(players)}
    for calling in callings:
        players_dict[players_dict_reverse[players_dict[calling] - 1]] += 1
        players_dict[calling] -= 1
        players_dict_reverse[players_dict[calling] + 1] = players_dict_reverse[players_dict[calling]]
        players_dict_reverse[players_dict[calling]] = calling
    return sorted(players_dict, key=lambda x: players_dict[x])

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;
class Solution {
    public String[] solution(String[] players, String[] callings) {
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < players.length; i++) {
            map.put(players[i], i);
        }
        for (String calling : callings) {
            int index = map.get(calling);
            // front는 앞에 있던 사람의 이름
            String front = players[index-1];
            players[index] = front;
            players[index-1] = calling;
            
            // map 업데이트
            map.put(calling, index-1);
            map.put(front, index);
        }
        return players;
    }
}